src/WebJobs.Extensions.DurableTask/Listener/DurableTaskTargetScaler.cs

```diff
@@ -0,0 +1,79 @@
+
+        public DurableTaskTargetScaler(string functionId, DurableTaskMetricsProvider metricsProvider, DurabilityProvider durabilityProvider, ILogger logger)
+        {
+            this.functionId = functionId;
+            this.metricsProvider = metricsProvider;
+            this.scaleResult = new TargetScalerResult();
+            this.TargetScalerDescriptor = new TargetScalerDescriptor(this.functionId);
+            this.durabilityProvider = durabilityProvider;
+            this.logger = logger;
+        }
+
+        public TargetScalerDescriptor TargetScalerDescriptor { get; private set; }
+
+        private int MaxConcurrentActivities => this.durabilityProvider.MaxConcurrentTaskActivityWorkItems;
+
+        private int MaxConcurrentOrchestrators => this.durabilityProvider.MaxConcurrentTaskOrchestrationWorkItems;
+
+        public async Task<TargetScalerResult> GetScaleResultAsync(TargetScalerContext context)
+        {
+            var metrics = await this.metricsProvider.GetMetricsAsync();
+
+            // compute activityWorkers: the number of workers we need to process all activity messages
+            var workItemQueueLength = metrics.WorkItemQueueLength;
+            double activityWorkers = Math.Ceiling(workItemQueueLength / (double)this.MaxConcurrentActivities);
+
+            var serializedControlQueueLengths = metrics.ControlQueueLengths;
+            var controlQueueLengths = JsonConvert.DeserializeObject<IReadOnlyList<int>>(serializedControlQueueLengths);
+
+            var controlQueueMessages = controlQueueLengths.Sum();
+            var activeControlQueues = controlQueueLengths.Count(x => x > 0);
+
+            // compute orchestratorWorkers: the number of workers we need to process all orchestrator messages.
+            // We bound this result to be no larger than the partition count
+            var upperBoundControlWorkers = Math.Ceiling(controlQueueMessages / (double)this.MaxConcurrentOrchestrators);
+            var orchestratorWorkers = Math.Min(activeControlQueues, upperBoundControlWorkers);
+
+            int numWorkersToRequest = (int)Math.Max(activityWorkers, orchestratorWorkers);
+            this.scaleResult.TargetWorkerCount = numWorkersToRequest;
+
+            // When running on ScaleController V3, ILogger logs are forwarded to Kusto.
+            var scaleControllerLog = $"Target worker count for {this.functionId}: {numWorkersToRequest}. " +
+                $"Metrics used: workItemQueueLength={workItemQueueLength}. controlQueueLengths={serializedControlQueueLengths}. " +
+                $"maxConcurrentOrchestrators={this.MaxConcurrentOrchestrators}. maxConcurrentActivities={this.MaxConcurrentActivities}";
+
+            // target worker count should never be negative
+            if (numWorkersToRequest < 0)
+            {
+                scaleControllerLog = "Tried to request a negative worker count." + scaleControllerLog;
+                this.logger.LogError(scaleControllerLog);
```

Comment:
Shouldn't we be writing structured logs that can be included in our Kusto tables?

Reply:
To be clear, this is not going to the Durable Functions Kusto table, but to the ScaleController table. It seems they want us to log using their ILogger implementation, and from there they have custom logic to forward that to Kusto. I'm just following their convention here.

If you'd like, I can _additionally_ log this data in our Kusto table via some structured log. I'd have to look whether or not we have the correct EventSource listener on the **ScaleController-host** to even do this though. After all, this code doesn't run in the context of the Functions runtime, but in the context of the Scale Controller process.